src/mame/sega/sega_beena.cpp

```diff
@@ -71,131 +102,2074 @@
+			m_card_data_i++;
+			if (m_card_data_i == 16) {
+				m_card_data_i = 0;
+				m_card_status = 0;
+				m_card_state = IDLE;
+			}
+		}
+
+		return 0x98 | (m_card_status << 6) | (data_bit << 5) | m_io_buttons->read();
+	}
+
+	return 0x98 | m_io_buttons->read();
+}
+
+void tvochken_state::install_game_rom()
+{
+	memory_region *rom = memregion("flash_rom");
+	m_maincpu->space(AS_PROGRAM).install_rom(ROM_FLASH_BASE, ROM_FLASH_BASE + 0x7fffff, 0x800000, rom->base());
 }
 
+
+static INPUT_PORTS_START( sega_9h0_0008 )
+	PORT_START("VIDEO_CONFIG")
+	PORT_CONFNAME( 0x80, 0x00, "Video Output" )
+	PORT_CONFSETTING( 0x00, DEF_STR( On ) )
+	PORT_CONFSETTING( 0x80, DEF_STR( Off ) )
+
+	PORT_START("PAGE_CONFIG")
+	PORT_CONFNAME( 0x80, 0x80, "Test Mode Pages" )
+	PORT_CONFSETTING( 0x00, DEF_STR( On ) )
+	PORT_CONFSETTING( 0x80, DEF_STR( Off ) )
+
+	PORT_START("CPU_CONFIG")
+	PORT_CONFNAME( 0x0f, 0x01, "IRQ Frequency" )
+	PORT_CONFSETTING( 0x01, "1 (Once every VBLANK)" )
+	PORT_CONFSETTING( 0x02, "2" )
+	PORT_CONFSETTING( 0x03, "3" )
+	PORT_CONFSETTING( 0x04, "4" )
+	PORT_CONFSETTING( 0x05, "5" )
+	PORT_CONFNAME( 0x10, 0x10, "IRQ Wait Speedup" )
+	PORT_CONFSETTING( 0x00, DEF_STR( On ) )
+	PORT_CONFSETTING( 0x10, DEF_STR( Off ) )
+	PORT_CONFNAME( 0x60, 0x20, "Memory Cache" )
+	PORT_CONFSETTING( 0x20, DEF_STR( On ) )
+	PORT_CONFSETTING( 0x40, DEF_STR( Off ) )
+INPUT_PORTS_END
+
+static INPUT_PORTS_START( sega_beena )
+	PORT_INCLUDE( sega_9h0_0008 )
+
+	PORT_START("PAD_LEFT")
+	PORT_BIT( 0x0001, IP_ACTIVE_LOW, IPT_JOYSTICKLEFT_UP ) PORT_8WAY PORT_PLAYER(1)
+	PORT_BIT( 0x0002, IP_ACTIVE_LOW, IPT_JOYSTICKLEFT_DOWN ) PORT_8WAY PORT_PLAYER(1)
+	PORT_BIT( 0x0004, IP_ACTIVE_LOW, IPT_JOYSTICKLEFT_LEFT ) PORT_8WAY PORT_PLAYER(1)
+	PORT_BIT( 0x0008, IP_ACTIVE_LOW, IPT_JOYSTICKLEFT_RIGHT ) PORT_8WAY PORT_PLAYER(1)
+	PORT_BIT( 0x0010, IP_ACTIVE_LOW, IPT_BUTTON1 ) PORT_PLAYER(1) PORT_NAME("Left Red Button")
+
+	PORT_START("PAD_RIGHT")
+	PORT_BIT( 0x0001, IP_ACTIVE_LOW, IPT_JOYSTICKRIGHT_UP ) PORT_8WAY PORT_PLAYER(1)
+	PORT_BIT( 0x0002, IP_ACTIVE_LOW, IPT_JOYSTICKRIGHT_DOWN ) PORT_8WAY PORT_PLAYER(1)
+	PORT_BIT( 0x0004, IP_ACTIVE_LOW, IPT_JOYSTICKRIGHT_LEFT ) PORT_8WAY PORT_PLAYER(1)
+	PORT_BIT( 0x0008, IP_ACTIVE_LOW, IPT_JOYSTICKRIGHT_RIGHT ) PORT_8WAY PORT_PLAYER(1)
+	PORT_BIT( 0x0010, IP_ACTIVE_LOW, IPT_BUTTON2 ) PORT_PLAYER(1) PORT_NAME("Right Red Button")
+
+	PORT_START("PEN_LEFT")
+	PORT_BIT( 0x01, IP_ACTIVE_HIGH, IPT_BUTTON3 ) PORT_PLAYER(1) PORT_NAME("Left Pen Button")
+	PORT_BIT( 0x02, IP_ACTIVE_HIGH, IPT_BUTTON4 ) PORT_PLAYER(1) PORT_NAME("Left Pen Target")
+
+	PORT_START("PEN_RIGHT")
+	PORT_BIT( 0x01, IP_ACTIVE_HIGH, IPT_BUTTON5 ) PORT_PLAYER(1) PORT_NAME("Right Pen Button")
+	PORT_BIT( 0x02, IP_ACTIVE_HIGH, IPT_BUTTON6 ) PORT_PLAYER(1) PORT_NAME("Right Pen Target")
+
+	PORT_START("PAGE")
+	PORT_BIT( 0x01, IP_ACTIVE_LOW, IPT_BUTTON7 ) PORT_PLAYER(1) PORT_NAME("Decrement Page")
+	PORT_BIT( 0x02, IP_ACTIVE_LOW, IPT_BUTTON8 ) PORT_PLAYER(1) PORT_NAME("Increment Page")
```

Comment:
Can you please use a positional input for the current page here as well?

Reply:
Nice, simplifies a bit of the parsing logic. I had to include a `PORT_CODE(JOYCODE_X)` just to avoid it mapping by default to the mouse, to avoid conflicts with the pen's lightgun mapping.